File: C++/23.合并-k-个升序链表.cpp
/*
 * @lc app=leetcode.cn id=23 lang=cpp
 *
 * [23] 合并 K 个升序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * struct ListNode {
 *     int val;
 *     ListNode *next;
 *     ListNode() : val(0), next(nullptr) {}
 *     ListNode(int x) : val(x), next(nullptr) {}
 *     ListNode(int x, ListNode *next) : val(x), next(next) {}
 * };
 */
class Solution {
public:
    ListNode* mergeKLists(vector<ListNode*>& lists) {
        return mergeSort(lists, 0, lists.size() - 1);
    }

    ListNode* mergeSort(vector<ListNode*>& lists, int start, int end) {
        if(start == end) return lists[start];
        if(start > end) return NULL;
        int mid = start + (end - start)/2;
        
        return mergeList(mergeSort(lists, start, mid), mergeSort(lists, mid+1, end));
    }

    ListNode* mergeList(ListNode* a, ListNode* b) {
        ListNode head, *tmp = &head, *l1 = a, *l2 = b;
        while(l1 && l2) {
            if(l1->val > l2->val) {
                tmp->next = l2;
                l2 = l2->next;
            } else {
                tmp->next = l1;
                l1 = l1->next;
            }
            tmp = tmp->next;
        }

        tmp->next = l1 ? l1 : l2;

        return head.next;
    }

};
// @lc code=end

File: Go/23.合并k个升序链表.go
/*
 * @lc app=leetcode.cn id=23 lang=golang
 *
 * [23] 合并K个升序链表
 */

// @lc code=start
/**
 * Definition for singly-linked list.
 * type ListNode struct {
 *     Val int
 *     Next *ListNode
 * }
 */
func mergeKLists(lists []*ListNode) *ListNode {
	return merge(lists, 0, len(lists)-1)
}

func merge(lists []*ListNode, start, end int) *ListNode {
	if start == end {
		return lists[start]
	}
	if start > end {
		return nil
	}
	mid := start + (end-start)/2
	return mergeTwoList(merge(lists, start, mid), merge(lists, mid+1, end))
}

func mergeTwoList(l1, l2 *ListNode) *ListNode {
	if l1 == nil && l2 == nil {
		return nil
	}
	tmp := &ListNode{}
	newList := tmp
	for l1 != nil && l2 != nil {
		if l1.Val < l2.Val {
			tmp.Next = l1
			l1 = l1.Next
		} else {
			tmp.Next = l2
			l2 = l2.Next
		}
		tmp = tmp.Next
	}
	if l1 == nil {
		tmp.Next = l2
	} else {
		tmp.Next = l1
	}
	return newList.Next
}

// @lc code=end

type ListNode struct {
	Val  int
	Next *ListNode
}

func main() {
	// 构建测试用的链表数组
	var lists []*ListNode
	for i := 0; i < 4; i++ {
		list := &ListNode{}
		curr := list
		for j := 1; j <= 5; j++ {
			curr.Val = i*5 + j
			if j < 5 {
				curr.Next = &ListNode{}
				curr = curr.Next
			}
		}
		lists = append(lists, list)
	}

	// 调用 mergeKLists 函数进行合并
	mergedList := mergeKLists(lists)

	// 输出合并后的链表结果
	printList(mergedList)
}
